GoRogue/MapGeneration/Region.cs

```diff
@@ -1,355 +1,478 @@
+            => new Region(r.MinExtent, (r.MaxExtentX, r.MinExtentY), r.MaxExtent, (r.MinExtentX, r.MaxExtentY));
+
+
+         /// <summary>
+         /// Creates a new Region in the shape of a parallelogram, with diagonals going down and right.
+         /// </summary>
+         /// <param name="origin">Origin of the parallelogram.</param>
+         /// <param name="width">Width of the parallelogram.</param>
+         /// <param name="height">Height of the parallelogram.</param>
+         /// <returns>A new region in the shape of a parallelogram</returns>
+         public static Region ParallelogramFromTopCorner(Point origin, int width, int height)
+         {
+             Point nw = origin;
+             Point ne = origin + new Point(width, 0);
+             Point se = origin + new Point(width * 2, height);
+             Point sw = origin + new Point(width, height);
+
+             return new Region(nw, ne, se, sw);;
+         }
+
+         /// <summary>
+         /// Creates a new Region in the shape of a parallelogram, with diagonals going up and right.
+         /// </summary>
+         /// <param name="origin">Origin of the parallelogram.</param>
+         /// <param name="width">The horizontal length of the top and bottom sides.</param>
+         /// <param name="height">Height of the parallelogram.</param>
+         /// <returns>A new region in the shape of a parallelogram</returns>
+         public static Region ParallelogramFromBottomCorner(Point origin, int width, int height)
+         {
+             var negative = Direction.YIncreasesUpward ? 1 : -1;
+
+             Point nw = origin + (height, height * negative);
+             Point ne = origin + (height + width, height * negative);
+             Point se = origin + (width, 0);
+             Point sw = origin;
+
+             return new Region(nw, ne, se, sw);
+         }
+        #endregion
+
+        /// <summary>
+        /// Returns a string detailing the region's corner locations.
+        /// </summary>
+        public override string ToString()
+            => $"Region: NW{NorthWestCorner.ToString()}=> NE{NorthEastCorner.ToString()}=> SE{SouthEastCorner.ToString()}=> SW{SouthWestCorner.ToString()}";
+
+        /// <summary>
+        /// Is this Point one of the corners of the Region?
+        /// </summary>
+        /// <param name="here">the point to evaluate</param>
+        public bool IsCorner(Point here) =>
+            here == NorthEastCorner || here == NorthWestCorner || here == SouthEastCorner || here == SouthWestCorner;
+
+
+        #region Transformation
+
+        /// <summary>
+        /// Moves the Region in the indicated direction.
+        /// </summary>
+        /// <param name="x">The X-value by which to shift the region</param>
+        /// <param name="y">The Y-value by which to shift the region</param>
+        /// <returns></returns>
+        public virtual Region Translate(int x, int y)
+            => Translate((x, y));
```

Comment:
Why virtual; is there a use-case for overriding?  Virtual comes with a performance cost, and potential dangers in forwarder implementations so I wanted to ask.

Also, minor consistency things:
- In library code I've tried to avoid implicit tuple conversions when it's similar difficulty to avoid; going x/y to tuple to Point instead of x/y to Point directly has some potential performance overhead.  It probably doesn't matter here because this code wouldn't really be used in a tight loop, but it might be worth adding anyway.

-  Similar function in `Rectangle` uses `dx` and `dy` as names for parameters

```suggestion
        public virtual Region Translate(int dx, int dy)
            => Translate(new Point(dx, dy));
```

Reply:
I was unsure whether we could implement this perfectly, and wanted to make it possible for users to extend it. An example use case would be to record a transformation in a user's extended method, but I think it's simpler to remove de-virtualize them, and re assess later